### PR TITLE
Field Search in Bulk Filter Modal

### PR DESCRIPTION
**Pull request by iethree**
## Description

Adds a search box that can be used to search for a field

fieldSearch

## To Test

- Search for fields in primary table
- Search for fields in linked tables
- Search for aggregate fields
- Search for "segments" or a segment name
- Search for something invalid so that you see the empty state
- Apply filters from the search view and make sure they show up properly when the search is cleared or applied
- type really fast... or slow to make sure the search debouncing works and the UI stays responsive

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx [hunk 1]
```diff
@@ -77,15 +94,17 @@ const BulkFilterModal = ({
...
         </ModalCloseButton>
       </ModalHeader>
-      {sections.length === 1 ? (
+      {sections.length === 1 || searchItems ? (
         <BulkFilterModalSection
           query={query}
           filters={filters}
-          section={sections[0]}
+          items={searchItems ?? sections[0].items}
+          isSearch={!!searchItems}
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx:107]: we need to pass down a prop indicating whether this is a search result so that the subcomponent can render the table name in addition to the field name

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx [hunk 2]
```diff
@@ -44,6 +55,12 @@ const BulkFilterModal = ({
     return query.topLevelFilterFieldOptionSections(null, 2, true);
   }, [query]);
 
+  const searchItems = useDebouncedFunction(
+    () => getSearchHits(searchQuery, sections),
+    200,
+    [searchQuery, sections],
+  );
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx:62]: there's some pretty significant processing overhead involved in filtering all these dimensions, so we need to debounce this.

**Comment by mazameli** [comment]
Nice! 

This whole double-x situation is bumming me out though. What if we just get rid of the outermost X on the modal, since you can close the modal with Cancel or esc?

image

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/utils.ts [hunk 3]
```diff
@@ -1,5 +1,7 @@
 import { DimensionOption } from "metabase-lib/lib/queries/StructuredQuery";
 
+import { isDate } from "metabase/lib/schema_metadata";
+
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/utils.ts:4]: I snuck this unrelated change into this branch to sort all date fields together

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.styled.tsx [hunk 4]
```diff
@@ -8,17 +10,18 @@ import IconButtonWrapper from "metabase/components/IconButtonWrapper";
 export const ModalRoot = styled.div`
   display: flex;
   flex-direction: column;
-  max-height: 90vh;
+  height: 90vh;
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.styled.tsx:13]: modal needs a fixed height now that we potentially have a rapidly varying number of fields on the screen

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.styled.tsx [hunk 5]
```diff
@@ -58,3 +62,17 @@ export const ModalCloseButton = styled(IconButtonWrapper)`
...
+  margin-right: ${space(2)};
+
+  input {
+    font-weight: bold;
+    padding: 12px ${space(3)};
+  }
+
+  .Icon-search {
+    margin-top: 2px;
+    transform: scale(0.8);
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.styled.tsx:76]: I hate this, but the search icon looks weird otherwise

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx [hunk 6]
```diff
@@ -77,15 +94,17 @@ const BulkFilterModal = ({
     <ModalRoot>
       <ModalHeader>
         <ModalTitle>{getTitle(question, query)}</ModalTitle>
+        <FieldSearch value={searchQuery} onChange={setSearchQuery} />
         <ModalCloseButton onClick={onClose}>
           <Icon name="close" />
         </ModalCloseButton>
       </ModalHeader>
-      {sections.length === 1 ? (
+      {sections.length === 1 || searchItems ? (
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/BulkFilterModal.tsx:102]: if the user is searching, we throw all the hits into a single section and hide the tabs

**Review by WiNloSt — APPROVED** [review]
Overall, this looks and functions pretty well. Good job on the new functionality 👍

**Comment by iethree** [comment]
> 

👍

File: frontend/src/metabase/hooks/use-debounced-function.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { useState, useEffect } from "react";
+
+export function useDebouncedFunction<TVALUE>(
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/hooks/use-debounced-function.ts:3]: maybe `useDebouncedEffect` would be more descriptive? I was trying to stay consistent with the useDebouncedValue, but pass a function instead of a value.

see: https://www.npmjs.com/package/use-debounced-effect

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/BulkFilterList.tsx [hunk 8]
```diff
@@ -47,6 +54,27 @@ const BulkFilterList = ({
     [options],
   );
 
+  if (!dimensions.length && !segments.length) {
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/BulkFilterList.tsx:57]: I really prefer the boolean expression and I think reads naturally is "does this have a length?" instead of making a comparison.  But if the other is more consistent with established patterns, that's fine by me.

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/utils.ts [hunk 9]
```diff
@@ -9,6 +11,7 @@ const fieldSortPriorities: PriorityMap = {
   "type/CreationTimestamp": 10,
   "type/CreationDate": 10,
   "type/CreationTime": 11,
+  "type/Date": 12,
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/utils.ts:14]: 😱

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts [hunk 10]
```diff
@@ -62,3 +68,37 @@ export const hasBackwardsArguments = (filter: Filter) => {
...
+  option?.dimension
+    ?.displayName()
+    ?.toLowerCase()
+    ?.includes(searchQuery?.toLowerCase());
+
+export const searchBySegmentName = (
+  segment: SegmentOption,
+  searchQuery: string,
+): boolean =>
+  "segments".includes(searchQuery?.toLowerCase()) ||
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: if there are segments defined for your table (your table probably doesn't have any), this will hit on the word segment in addition to the user-defined names of the segments

Screen Shot 2022-07-11 at 12 20 49 PM

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts [hunk 11]
```diff
@@ -62,3 +68,37 @@ export const hasBackwardsArguments = (filter: Filter) => {
...
+  searchQuery: string,
+  sections: FilterSection[],
+) => {
+  if (searchQuery === "") {
+    return null;
+  }
+
+  return sections
+    .map(s => s.items)
+    .flat()
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:98]: cool! didn't know that was in vanilla js!

File: frontend/src/metabase/hooks/use-debounced-function.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/hooks/use-debounced-function.ts]: I see what you're trying to do here. I didn't realize we have https://github.com/metabase/metabase/blob/b5747678f4d3381dc803edabbebf97377e30126e/frontend/src/metabase/hooks/use-debounced-value.ts#L3

I'd just extend that hook and accept functions that produce value in addition to just accepting a value. There are a bunch of API out there that do this kind of thing, either pass a value or a function that produce value. e.g `setState` you can pass either a value or a function`. What do you think?

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: Hah, I see! Do you think the segment in the search should also say what table it's from like other searched attributes?

image
vs
image


Also, I'm not totally sure how segment searching works. I can select segment when viewing **People** and **Order** individually
image

image

But if I want to filter **User** (**Person** table) from **Order** then I won't see Person segment anymore.
image

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/BulkFilterList.styled.tsx [hunk 12]
```diff
@@ -33,3 +33,7 @@ export const FilterDivider = styled.div`
     display: none;
   }
 `;
+
+export const EmptyStateText = styled.span`
```
Comment (WiNloSt) [inline frontend/src/metabase/query_builder/components/filters/modals/BulkFilterList/BulkFilterList.styled.tsx:37]: I think this is safe to delete now right?

File: frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: Having a constant on the left feels unusual. The way I read that is _"string 'segments' includes a search query"_, however, it's supposed to be something like _"search query contains the 'segments' string"_ I believe? I'd expect it to be something like `searchQuery?.toLowerCase().includes("segments")`. This is totally nit though

[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: @kulyk  except that won't match on the substring "segme" ??  i think the format is consistent with the format search `haystack.includes(needle)`

[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: Oh, that makes sense, thanks for explaining 👍

[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/query_builder/components/filters/modals/BulkFilterModal/utils.ts:85]: @WiNloSt, segments should be table specific - one dropdown for a set of segments for each table. good catch on the table names. I'll make sure those show up.
